src/main/java/mate/academy/dao/impl/MovieSessionDaoImpl.java

```diff
@@ -0,0 +1,78 @@
+import java.util.Optional;
+import java.util.stream.Collectors;
+import mate.academy.dao.MovieSessionDao;
+import mate.academy.exception.DataProcessingException;
+import mate.academy.lib.Dao;
+import mate.academy.model.MovieSession;
+import mate.academy.util.HibernateUtil;
+import org.hibernate.Session;
+import org.hibernate.Transaction;
+import org.hibernate.query.Query;
+
+@Dao
+public class MovieSessionDaoImpl implements MovieSessionDao {
+
+    @Override
+    public MovieSession add(MovieSession movieSession) {
+        Transaction transaction = null;
+        Session session = null;
+        try {
+            session = HibernateUtil.getSessionFactory().openSession();
+            transaction = session.beginTransaction();
+            session.save(movieSession);
+            transaction.commit();
+            return movieSession;
+        } catch (Exception e) {
+            if (transaction != null) {
+                transaction.rollback();
+            }
+            throw new DataProcessingException("Can't insert movie session " + movieSession, e);
+        } finally {
+            if (session != null) {
+                session.close();
+            }
+        }
+    }
+
+    @Override
+    public Optional<MovieSession> get(Long id) {
+        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
+            Query<MovieSession> query = session.createQuery("from MovieSession m "
+                    + "left join fetch m.movie "
+                    + "left join fetch m.cinemaHall "
+                    + "where m.id = :id", MovieSession.class);
+            query.setParameter("id", id);
+            return Optional.ofNullable(query.getSingleResult());
+        } catch (Exception e) {
+            throw new DataProcessingException("Can't get a movie session by id: " + id, e);
+        }
+    }
+
+    @Override
+    public List<MovieSession> findAvailableSessions(Long movieId, LocalDate date) {
+        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
+            Query<MovieSession> query = session.createQuery("from MovieSession ms "
+                            + "left join fetch ms.movie "
+                            + "left join fetch ms.cinemaHall "
+                            + "WHERE ms.movie.id = :movieId "
+                            + "AND YEAR(ms.showTime) = :year "
+                            + "AND MONTH(ms.showTime) = :month "
+                            + "AND DAY(ms.showTime) = :day",
+                    MovieSession.class);
+            query.setParameter("movieId", movieId);
+            query.setParameter("year", date.getYear());
+            query.setParameter("month", date.getMonthValue());
+            query.setParameter("day", date.getDayOfMonth());
+            return query.getResultList()
+                    .stream()
+                    .filter(m -> m.getShowTime().getDayOfYear() == date.getDayOfYear())
+                    .collect(Collectors.toList());
```

Comment:
in my opinion, this is unnecessary filtering, after we set the selection conditions when generating query parameters

Reply:
+, it is in fact a bad practice to filter data after it was already fetched from db, we have to filter it with the SQL queries, because it is the main bottleneck of the whole application